src/SpectatorView.Unity/Assets/SpectatorView/Scripts/StateSynchronization/AssetCache.cs

```diff
@@ -277,35 +404,129 @@ public override void ClearAssetCache()
+            Dictionary<string, List<AssetId>> oldAssets = lookupByName == null ? new Dictionary<string, List<AssetId>>() : lookupByName;
+            HashSet<Tuple<string, AssetId>> unvisitedIds = new HashSet<Tuple<string, AssetId>>();
+            lookupByAsset = new Dictionary<TAsset, AssetCacheEntry>();
+            lookupByAssetId = new Dictionary<AssetId, AssetCacheEntry>();
+            foreach (var listPair in oldAssets)
+            {
+                foreach (var id in listPair.Value)
+                {
+                    unvisitedIds.Add(new Tuple<string, AssetId>(listPair.Key, id));
+                }
+            }
 
             foreach (TAsset asset in EnumerateAllAssets())
             {
-                unvisitedAssets.Remove(asset);
-                if (!oldAssets.ContainsKey(asset))
+                if (asset.name == null ||
+                    asset.name == string.Empty ||
+                    !AssetDatabase.TryGetGUIDAndLocalFileIdentifier(asset, out string guid, out long localid))
+                {
+                    Debug.LogError($"Unable to identify asset: {asset}");
+                    continue;
+                }
+
+                var assetId = new AssetId(new Guid(guid), localid, asset.name);
+                unvisitedIds.Remove(new Tuple<string, AssetId>(assetId.Name, assetId));
+
+                if (!oldAssets.ContainsKey(assetId.Name))
+                {
+                    oldAssets[assetId.Name] = new List<AssetId>();
+                    oldAssets[assetId.Name].Add(assetId);
+                }
+                else if (!oldAssets[assetId.Name].Contains(assetId))
+                {
+                    oldAssets[assetId.Name].Add(assetId);
+                }
+
+                var entry = new AssetCacheEntry();
+                entry.Asset = asset;
+                entry.AssetId = assetId;
+                lookupByAsset[asset] = entry;
+                lookupByAssetId[assetId] = entry;
+            }
+
+            foreach (var idPair in unvisitedIds)
+            {
+                if (oldAssets.TryGetValue(idPair.Item1, out List<AssetId> ids) &&
+                    ids.Contains(idPair.Item2))
                 {
-                    if (AssetDatabase.TryGetGUIDAndLocalFileIdentifier(asset, out string guid, out long localid))
+                    ids.Remove(idPair.Item2);
+                    if (ids.Count == 0)
                     {
-                        oldAssets.Add(asset, new TAssetEntry
-                        {
-                            AssetId = new AssetId(new Guid(guid), localid),
-                            Asset = asset
-                        });
+                        string assetName = GetAssetCachesContentPath($"{GetValidAssetName(idPair.Item1)}_{this.GetType().Name}", assetContentExtension);
+                        AssetDatabase.DeleteAsset(assetName);
                     }

```

Comment:
>} [](start = 20, length = 1)

If the count of the ID list is now zero, should we also simply remove the entry from oldAssets entirely? i.e. any reason to store an empty list?

Reply:
good catch

---
In reply to: [355047556](https://github.com/microsoft/MixedReality-SpectatorView/pull/295#discussion_r355047556) [](ancestors = 355047556)